### PR TITLE
Gutenboarding: add domain availability check

### LINKED ISSUE #41220: Gutenboarding: before domain is confirmed, check that it really is available
via https://github.com/Automattic/wp-calypso/issues/40971#issuecomment-614828121

> The moment customer picks a domain - basically you have to assume that the suggestion endpoint is not always 100% accurate and might return unavailable domains. So, once a domain is selected, you should for it's availability via the is-available endpoint.
You can see it in action in the current domain search - at first I select a suggestion that turns out to unavailable and then one that's available and I'm allowed to proceed.
>
> https://cloudup.com/c0yynmBzeud

Some existing code:

https://github.com/Automattic/wp-calypso/blob/711190aee23edea9006c079d44b7b7cee5d9b977/client/lib/wpcom-undocumented/lib/undocumented.js#L482-L492

https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/domains/registration/availability-messages.js

See details in pbAok1-Hb-p2#comment-1472

Related: https://github.com/Automattic/wp-calypso/issues/41219

## Design

Design for state when domain isn't available:

image

@dubielzyk how should the "loading" state look like after selecting a domain from the list? I don't know how long it takes typically but we should make sure it doesn't look bad on slow connections or if API is slow to respond.

We can meanwhile implement this without the loading state.

### PR DESCRIPTION
**Pull request by sirreal**
#### Background

This check is being split into two separate PRs.

1. Setup the selectors for the domain availability check.
2. Use these selectors on the UI.

The second PR hasn’t been opened yet but these selectors would be used in the following scenarios:

**Scenario 1:**
A user enters a FQDN for their site name. E.g.

* “Automattic” - Check won’t be performed.
* “Automattic.com” - Check will be performed.

**Scenario 2:**
A user selects a domain that is displayed as a suggested domain.

**Scenario 3:**
In the case the user types a FQDN in the search bar on `/new/domains`, we’d check the availability of that domain specifically as soon as the user finished typing

There’s more on this here https://github.com/Automattic/wp-calypso/issues/40971#issuecomment-614828121

> Basically you have to assume that the suggestion endpoint is not always 100% accurate and might return unavailable domains. So, once a domain is selected, you should for it’s availability via the is-available endpoint.

#### Changes proposed in this Pull Request

* Adding domain availability selector to the `domain-suggestions` package.
* If the API request fails, we will return a status of `'unknown'` which is what the API does if it can not accurately determine the availability status.

#### Testing instructions

These changes are not yet being used, so we would be best checking the absence of them and that everything else around domain suggestions runs as normal.

* Go to [/new](https://hash-fb5de047678ac106a94e2d36f47131b5ac43f203.calypso.live/new) and enter a site name
* Check the network tab that a request is not being made to the `/is-available` API
* Check that the domain suggestions get returned as normal and you've able to proceed through the process.

Partially implements changes requested by #41220

### CODE REVIEW

**Comment by sirreal** [comment]
This could be landed with few or no changes, just adding the selector and implementation to the data-stores package.

File: packages/data-stores/src/domain-suggestions/resolvers.ts
Comment (sirreal) [reply packages/data-stores/src/domain-suggestions/resolvers.ts:24]: I didn't change the signature of `fetchAndParse` here becuase I was expecting it to come from https://github.com/Automattic/wp-calypso/pull/41450

Depends on which lands first 🤷‍♂️

File: packages/data-stores/src/domain-suggestions/resolvers.ts
Comment (sirreal) [inline packages/data-stores/src/domain-suggestions/resolvers.ts:15]: It's a bit weird to use `TailParameters<…>[0]` and `Parameters<…>[1]` (effectively the same) in the same file, but it gets the job done 🙂

**Review by sirreal — COMMENTED** [review]
The changes seem pretty reasonable to me. I haven't tested, I'll leave this to @Automattic/luna to get this landed.

**Review by ciampo — CHANGES_REQUESTED** [review]
Code changes look good to me!

I would probably update the PR description to add some more context:
- We could clarify that a second PR (making use of the validation check) will be opened.
- We could also change the "Fixes #41220" line to something like "Partially implements changes requested by #41220", just to avoid closing that issue when this PR gets merged

Also, is error handling going to be handled in a separate PR (see also this comment https://github.com/Automattic/wp-calypso/issues/44421#issuecomment-682589700)
I already started working on error handling in #45468, although it's still far from being ready.

File: packages/data-stores/src/domain-suggestions/resolvers.ts
Comment (ciampo) [inline packages/data-stores/src/domain-suggestions/resolvers.ts:26]: Out of curiosity, do we already know how this will be addressed in the UI / from the point of view of the user?

**Review by ciampo — APPROVED** [review]
Changes LGTM! The PR description is very clear and gives context to this set of changes 😃

I just a left a comment, but that's more a clarification that doesn't impact this PR.

Let's wait for reviews from the rest of the team

**Review by alshakero — APPROVED** [review]
Code LGTM!

File: packages/data-stores/src/domain-suggestions/resolvers.ts
Comment (alshakero) [reply packages/data-stores/src/domain-suggestions/resolvers.ts:26]: Not yet. This functionality is yet to be used.

**Pull request merged**
